riscv/debug_module.cc

```diff
@@ -715,120 +730,120 @@ bool debug_module_t::perform_abstract_command()
-              break;
-            case 3:
-              write32(debug_abstract, i++, ld(S0, ZERO, debug_data_start));
-              break;
-            default:
-              abstractcs.cmderr = CMDERR_NOTSUP;
-              return true;
-          }
-          write32(debug_abstract, i++, csrw(S0, regno));
-
-        } else {
-          write32(debug_abstract, i++, csrr(S0, regno));
-          switch (size) {
-            case 2:
-              write32(debug_abstract, i++, sw(S0, ZERO, debug_data_start));
-              break;
-            case 3:
-              write32(debug_abstract, i++, sd(S0, ZERO, debug_data_start));
-              break;
-            default:
-              abstractcs.cmderr = CMDERR_NOTSUP;
-              return true;
-          }
-        }
-        if (!is_fpu_reg(regno)) {
-          write32(debug_abstract, i++, csrr(S0, CSR_DSCRATCH0));
+      if (write) {
+        switch (size) {
+          case 2:
+            write32(debug_abstract, i++, lw(S0, ZERO, debug_data_start));
+            break;
+          case 3:
+            write32(debug_abstract, i++, ld(S0, ZERO, debug_data_start));
+            break;
+          default:
+            abstractcs.cmderr = CMDERR_NOTSUP;
+            return true;
         }
+        write32(debug_abstract, i++, csrw(S0, regno));
 
-      } else if (regno >= 0x1000 && regno < 0x1020) {
-        unsigned regnum = regno - 0x1000;
-
+      } else {
+        write32(debug_abstract, i++, csrr(S0, regno));
         switch (size) {
           case 2:
-            if (write)
-              write32(debug_abstract, i++, lw(regnum, ZERO, debug_data_start));
-            else
-              write32(debug_abstract, i++, sw(regnum, ZERO, debug_data_start));
+            write32(debug_abstract, i++, sw(S0, ZERO, debug_data_start));
             break;
           case 3:
-            if (write)
-              write32(debug_abstract, i++, ld(regnum, ZERO, debug_data_start));
-            else
-              write32(debug_abstract, i++, sd(regnum, ZERO, debug_data_start));
+            write32(debug_abstract, i++, sd(S0, ZERO, debug_data_start));
             break;
           default:
             abstractcs.cmderr = CMDERR_NOTSUP;
             return true;
         }
+      }
+      if (!is_fpu_reg(regno)) {
+        write32(debug_abstract, i++, csrr(S0, CSR_DSCRATCH0));
+      }
 
-        if (regno == 0x1000 + S0 && write) {
-          /*
-           * The exception handler starts out be restoring dscratch to s0,
-           * which was saved before executing the abstract memory region. Since
-           * we just wrote s0, also make sure to write that same value to
-           * dscratch in case an exception occurs in a program buffer that
-           * might be executed later.
-           */
-          write32(debug_abstract, i++, csrw(S0, CSR_DSCRATCH0));
-        }
+    } else if (regno >= 0x1000 && regno < 0x1020) {
```

Comment:
```suggestion
    } else if (regno < 0x1020) {
```

It’s just a personal preference, but I find it easier to read and compare against the spec if less than or equal (regno <= 0x101f) is used.